cocotb/share/lib/embed/gpi_embed.cpp

```diff
@@ -188,172 +188,138 @@ extern "C" COCOTB_EXPORT void _embed_sim_cleanup(void) {
+    if (!filter_func) {
         // LCOV_EXCL_START
         PyErr_Print();
-        LOG_ERROR("Failed to get the _sim_event method");
-        goto cleanup;
+        return -1;
         // LCOV_EXCL_STOP
     }
+    DEFER(Py_DECREF(filter_func));
 
-    cocotb_init = PyObject_GetAttrString(
-        cocotb_module, "_initialise_testbench");  // New reference
-    if (cocotb_init == NULL) {
+    py_gpi_logger_initialize(log_func, filter_func);
+
+    pEventFn = PyObject_GetAttrString(entry_module, "_sim_event");
+    if (!pEventFn) {
         // LCOV_EXCL_START
         PyErr_Print();
-        LOG_ERROR("Failed to get the _initialise_testbench method");
-        goto cleanup;
+        return -1;
         // LCOV_EXCL_STOP
     }
+    // cocotb must hold _sim_event until _embed_sim_end runs
 
     // Build argv for cocotb module
-    argv_list = PyList_New(argc);  // New reference
+    auto argv_list = PyList_New(argc);
     if (argv_list == NULL) {
         // LCOV_EXCL_START
         PyErr_Print();
-        LOG_ERROR("Unable to create argv list");
-        goto cleanup;
+        return -1;
         // LCOV_EXCL_STOP
     }
-    for (i = 0; i < argc; i++) {
+    for (int i = 0; i < argc; i++) {
         // Decode, embedding non-decodable bytes using PEP-383. This can only
         // fail with MemoryError or similar.
-        PyObject *argv_item = PyUnicode_DecodeLocale(
-            argv[i], "surrogateescape");  // New reference
-        if (argv_item == NULL) {
+        auto argv_item = PyUnicode_DecodeLocale(argv[i], "surrogateescape");
+        if (!argv_item) {
             // LCOV_EXCL_START
             PyErr_Print();
-            LOG_ERROR(
-                "Unable to convert command line argument %d to Unicode string.",
-                i);
-            Py_DECREF(argv_list);
-            goto cleanup;
+            return -1;
             // LCOV_EXCL_STOP
         }
-        PyList_SET_ITEM(argv_list, i, argv_item);  // Note: This function steals
-                                                   // the reference to argv_item
+        PyList_SetItem(argv_list, i, argv_item);
     }
+    DEFER(Py_DECREF(argv_list))
 
-    cocotb_retval = PyObject_CallFunctionObjArgs(cocotb_init, argv_list, NULL);
-    Py_DECREF(argv_list);
-    Py_DECREF(cocotb_init);
-
-    if (cocotb_retval != NULL) {
-        LOG_DEBUG("_initialise_testbench successful");
-        Py_DECREF(cocotb_retval);
-    } else {
+    auto cocotb_retval =
+        PyObject_CallFunctionObjArgs(entry_point, argv_list, NULL);
+    if (!cocotb_retval) {
         // LCOV_EXCL_START
         PyErr_Print();
-        LOG_ERROR("cocotb initialization failed - exiting");
-        goto cleanup;
+        return -1;
         // LCOV_EXCL_STOP
     }
+    Py_DECREF(cocotb_retval);
 
-    goto ok;
-
-cleanup:
-    ret = -1;
-ok:
-    Py_XDECREF(cocotb_module);
-    Py_XDECREF(cocotb_log_module);
+    LOG_INFO("cocotb initialization successful");
```

Comment:
Should this stay `DEBUG`? Or maybe removed altogether. It doesn't show up until the first yield back to the simulator, so it can look out of place. 

Example: https://github.com/cocotb/cocotb/pull/2484/checks?check_run_id=2157206436#step:25:4296

Reply:
I thought the same thing. I'd like to remove it.
```suggestion
```